### ISSUE TITLE
[NOT AN ISSUE] Useful DMA Info

### ISSUE THREAD

**Issue by deathbybandaid** [wouterdebie/locast2dvr#26]
I tossed together a quick python script that checks all possible DMA numbers on Locast for availability, and outputs how many channels are available in that DMA. Here's a paste of the results for those interested

https://pastebin.com/C6ujdwzi

**Comment by wouterdebie** [comment]
Great! Are you using the locast.org API (https://api.locastnet.org/api/dma) for this? How did you get the channel count?

**Comment by deathbybandaid** [comment]
yes I am! 

https://github.com/fHDHR/fHDHR_plugin_origin_locast/blob/main/web/locast_api_tools.py#L76-L101

https://github.com/fHDHR/fHDHR_plugin_origin_locast/blob/main/web/dma_markets.py

**Comment by wouterdebie** [comment]
Great, thanks! Where do you use this for?

**Comment by deathbybandaid** [comment]
I'm the dev of fHDHR. I based my code on locast2plex, but development was too slow, and now my code doesn't even resemble l2p.

I made fHDHR into a plugin-based system, to allow different 'origin' media like locast, PlutoTV, tvheadend, etc. I added a database, a web interface,and removed the requirement of ffmpeg.

https://github.com/fHDHR/fHDHR

**Comment by wouterdebie** [comment]
Yeah, I figured :) Thanks for the heads-up!

**Comment by Croq360** [comment]
@deathbybandaid - I'm intrigued by the "removed the requirement of ffmpeg" statement.  In some testing with multiple recordings at once using Plex, that seemed to be the bottleneck for me, and it would start falling behind.  Sometimes with certain programs, I'd even get missing segments when recording just one thing at once.

**Comment by deathbybandaid** [comment]
I did a bunch of research on m3u8, and wrote a method to do this with just python requests module. I also handle decryption for encrypted m3u8 streams.

Recently I even wrote a basic quality selection setting for high,medium,low.

Ffmpeg and VLC are also stream plugins for those that prefer.

**Comment by wouterdebie** [comment]
Really cool! I just looked at your code and it's fairly straight forward. @Croq360 I'll some functionality to remove ffmpeg as well, so you can test if you want. 
That said, I just got rate limited by locast, since I requested the steam uri way too often because of a bug :) But I'll continue when the rate limiting is over :)

**Comment by Croq360** [comment]
Sure, happy to test out!  Just let me know.  Maybe I'll dust off my alternate remapper (by callsign) too if we can get that going.

What's the "rate limiting"?  Is that when you've watched too many channels too quickly?

**Comment by wouterdebie** [comment]
@Croq360 I released this in 0.7.0. There's a new option `-ds/--direct` that you can use to enable this. 
@deathbybandaid thanks for this! I took inspiration from your code. One thing I added was timing of serving chunks. Locast2dvr uses Flask and waitress to serve and it seems that output is buffered, so yielding chunks is pretty much instant. This caused locast2dvr hammering the locast stream_url, so I implemented a way to throttle that. I commented it here: https://github.com/wouterdebie/locast2dvr/blob/6a7990129616cd93fea7d316bdf816aa7e6d7301/locast2dvr/http/interface.py#L463

**Comment by wouterdebie** [comment]
> What's the "rate limiting"? Is that when you've watched too many channels too quickly?

Yes, locast rates the amount of requests you can do per some timeframe (minutes, seconds). I made a mistake and my code kept on hammering their API. On next login I got blocked. Turns out you have to wait an hour for it to go away though.

**Comment by deathbybandaid** [comment]
I'll check that out. When I wrote my iteration, I did find I was hitting the m3u8 more often than needbe.

I will probably steal some code from you! I didn't know about the `.duration` attribute for m3u8 chunks.

**Comment by deathbybandaid** [comment]
I really like your `OrderedDict` system over my current list system. Really slick!

**Comment by deathbybandaid** [comment]
One thing you don't have in your version is decrypting the chunks, which isn't needed for locast streams, but fHDHR supports services like PlutoTV that has to decrypt.

One thing I REALLY need to figure out is adding an abstraction layer to allow fHDHR to serve multiple clients from the same stream, which would really help with the locast stream limitations.

**Comment by deathbybandaid** [comment]
One tip I'd have for you that seems to help, is making use of `request.session`, seems to help with locast especially.

**Comment by deathbybandaid** [comment]
I'll also leave the below link here, if you are interested in also supporting VLC streaming:

https://github.com/fHDHR/fHDHR_plugin_stream_vlc/blob/main/__init__.py

**Comment by Croq360** [comment]
@wouterdebie - I'll check it out.  Should I wait for you to look at the "request.session" thing that @deathbybandaid or is it worth testing out anyway without that?  There is some time on my end as I can't test until a bit later tonight (I have one thing recording early, so can set up some other simultaneous recordings after that to see how they do).

Regarding the "rate limit" thing, I even ran into that once or twice with just normal watching (manually selecting channels).  I was channel surfing, going to a channel, watching for a few seconds, then to something else.  Got the message about it, had to wait 30 minutes or an hour before I could watch anything else.  The lousy thing about it (though I think I can see why they do it) is that if you happen to try again say 29 minutes in, your timer till it unlocks resets.

**Comment by wouterdebie** [comment]
> I will probably steal some code from you! 
Please do!

> I didn't know about the `.duration` attribute for m3u8 chunks.
I didn't know either :) I went down a rabbit hole of parsing the .ts stream to figure out the duration and/or bitrate, but this was much easier :)

> I really like your `OrderedDict` system over my current list system. Really slick!
Thanks! I like clean code :)

> One thing you don't have in your version is decrypting the chunks, which isn't needed for locast streams, but fHDHR supports services like PlutoTV that has to decrypt.
I deliberately didn't implement this, since locast.org doesn't encrypt strings, so it would be kind of pointless. I like to create things I use myself and I'm not using anything else than locast right now, but if if I do, I'll have a look.
 
> One thing I REALLY need to figure out is adding an abstraction layer to allow fHDHR to serve multiple clients from the same stream, which would really help with the locast stream limitations.
I'd probably start a thread that reads data from locast (as you do now) and then send it to a queue. Then have clients read from that queue. The challenging part is probably some synchronization between producer and consumer and when to invalidate stuff from the queue when there are multiple consumers.
It's interesting that you mention this, because it might be fun to implement this though.. And I might :)

> One tip I'd have for you that seems to help, is making use of `request.session`, seems to help with locast especially.
I'm aware. Most calls can don't have to be authenticated and I'd rather keep it that way, since it'd be harder to piece together requests on the locast.org side. One thing I'd like to do is to refactor the use of `requests` into a more central place, because it's sprinkled all over the place now.

> I'll also leave the below link here, if you are interested in also supporting VLC streaming:
Thanks for this. Not right now. If streaming with out ffmpeg works, I'm inclined to rip it out completely, since there's no real reason for doing so. I rather support less features than more :)

I saw you built in quality selection as well. Is this something that is supported by Emby or PMS? Right now I just pick the stream with the highest quality and use that.

**Comment by deathbybandaid** [comment]
The quality setting is just in the config settings.

However, I worked really hard to emulate what a real hdhr does

**Comment by deathbybandaid** [comment]
@wouterdebie Below might be useful to you for quality selection. This has logic to deal with multiple nested playlists, and m3u8 with variant streams not in any particular order. I also built in some logic to re-sort the variants by bandwidth and resolution (resolution isn't always given.) I have 3 quality settings `high,medium,low` which just select `-1,len/2,0`. This isn't perfect, but different m3u8 streams will have different amounts of variant options.


https://github.com/fHDHR/fHDHR/blob/b81752864ee270313acfd881000aa6a1c71a34e5/fHDHR/device/tuners/__init__.py#L180-L258

**Comment by Croq360** [comment]
@wouterdebie - I did some testing in PMS last night, and it seemed mostly okay.  One of the shows had a part where it jumped forward from some missing video for maybe 1-3 seconds (maybe it was a dropped segment or something?).  I'll continue watching the other shows to see if I notice any other issues.

I also noticed that the length of the recordings are a few seconds short of being full length.  I start my recordings 1 minute early and end 3 minutes late, and instead of being 01:04:00 long, they all seem to be 01:03:57 or 01:03:56.

All that being said, this new direct streaming seems much more stable than what we were seeing when using ffmpeg.

I'll continue more testing today.  I know there were certain programs that were always having issues before and causing multiple retries, and I'll monitor those as well.

**Comment by wouterdebie** [comment]
@Croq360 thanks for testing! I'm pretty sure that the time issue doesn't have to do with locast2dvr, since locast2dvr doesn't have a notion of program. Plex/Emby ask for a stream and stop whenever they want. The request for the stream happens only once and locast2dvr just serves continuously as long as Plex/Emby doesn't disconnect.

The 1-3 seconds jumps are interesting, since I'd expect at least a 10 second skip. Locast provides 10 second segments, so if one gets dropped, the skip should be at least that long. Skipping a partial segment in locast2dvr is unlikely. If Plex/Emby would have disconnected (something that should be visible in the locast2dvr logs), it could be possible that the final output has some jumps backwards or forwards in time, however still in multiples of 10 seconds.
How often did this happen? Just once? Technically it could have happened when locast.org itself recorded the actual stream.

**Comment by deathbybandaid** [comment]
I've been using my version of the "direct" method for a long time now, and I think it works just as well.

**Comment by Croq360** [comment]
So far, in my testing, the direct method is working MUCH better than ffmpeg did for Plex.  I have more stuff to record today and will set up some other record-once shows to keep testing.

Regarding the lengths of the recordings, they are all showing as 3-4 seconds short of the full length.  I'll take a look at the logs and see if this is just Plex starting and/or ending things a second or two off of the scheduled time.  This is not a big deal though, especially with my pre- and post- time buffers.

I'll also look at the logs to see if there's anything in there about the missing segments (it could have been 10 seconds, hard to tell exactly).  It seems that the segment is in the video, given the length, but it's skipping over that piece for some reason.

**Comment by wouterdebie** [comment]
>  I also built in some logic to re-sort the variants by bandwidth and resolution (resolution isn't always given.)

It's similar to what I do, however, I just pick the top resolution and do that in a one-liner :) If there are users of locast2dvr that always want a specific max resolution, I'll implement it, but there is no real need for now. In the locast case, the m3u8 from the stream_url either contains a single url to a stream, in which case I just use that, or a list of playlists with specified resolution. And since resolution and bandwith are related, I just sort on resolution (sorting tuples seems to only sort on the first element, but that is fine in this case, because we're always dealing with a screen width that is larger than the screen height).

One thing I noticed though is that you keep quite a lot of counters e.g in https://github.com/fHDHR/fHDHR/blob/b81752864ee270313acfd881000aa6a1c71a34e5/fHDHR/device/tuners/__init__.py#L213, while you could just use `for playlist_index, playlist_item in enumerate(videoUrlM3u.playlists):`.

**Comment by wouterdebie** [comment]
> see if this is just Plex starting and/or ending things a second or two off of the scheduled time.

This would be my guess. it could very well be that some buffers go unflushed, or that the last segment is streamed, but not written to disk. That said, it's good to know that the overall experience is much better. I'll add some additional logging for skipped segments later today.

**Comment by wouterdebie** [comment]
> I've been using my version of the "direct" method for a long time now, and I think it works just as well.

What use cases do you have for continue to support ffmpeg and vlc?

**Comment by deathbybandaid** [comment]
Yup, always improving my codebase as I learn, I'll look at my codebase for indexing next.

I have some users that claim vlc/ffmpeg methods as being superior for their usecases, but I've moved vlc/ffmpeg to their own plugins, seperate from the core code.

One thing on my long-term list is to support hardware transcoding as well.

I also have a plugin that adds the ability to stream in the web interface (uses ffmpeg).

https://github.com/fHDHR/fHDHR_plugin_web_watch

**Comment by deathbybandaid** [comment]
I also didn't know exactly how emby handled "m3u tuners", so I'm also now implementing a method for fHDHR to redirect to the m3u8 url directly

**Comment by wouterdebie** [comment]
@deathbybandaid ah gotcha.. The philosophy behind locast2dvr (and that's not better or worse than any thing else) is to keep it as simple as possible and don't have feature creep. I'll let transcoding and stuff handled by Emby/PMS.

**Comment by wouterdebie** [comment]
> I also didn't know exactly how emby handled "m3u tuners", so I'm also now implementing a method for fHDHR to redirect to the m3u8 url directly

The only thing I do is redirect (http 302) to a locast m3u url. Unfortunately Emby doesn't support playlists with multiple resolutions (yet), so I have to still get the playlist with a specific resolution.

**Comment by deathbybandaid** [comment]
I work in IT, and between calls, I code. I code out of boredom, feature creep is real, but it's always one more hurtle challenge to keep my mind active.

**Comment by deathbybandaid** [comment]
The only catch I'm seeing with implementing that redirect in fHDHR, is controlling the amount of client streams

**Comment by deathbybandaid** [comment]
@wouterdebie I'm going to shoot you an email.

**Comment by Croq360** [comment]
@wouterdebie - I just had a show record that is missing over half of the video when it finished.  I'll grab your latest update with the additional logging on segments and try it again in a different timeslot.

By the way, in the log currently, there were a ton of messages (dozens per second when they happened) where it stated:

INFO - HTTPInterface: Added 0 new segments, removed 0

Maybe that can give a clue?

**Comment by wouterdebie** [comment]
@Croq360 that is indeed a good clue.. this means that all know segments have been sent to the client, but there were no new ones. Do you have the full log?

**Comment by Croq360** [comment]
Unfortunately, no, I already upgraded to the latest version.  If it happens again, I'll capture it and send it along.

**Comment by wouterdebie** [comment]
I've added even more logging in 0.7.3. Make sure you run with -v/--verbose

**Comment by wouterdebie** [comment]
> but I was watching something on Emby at the same time

Technically it shouldn't, but you never know :)

**Comment by wouterdebie** [comment]
> The only catch I'm seeing with implementing that redirect in fHDHR, is controlling the amount of client streams

I hear you, but why would you care?

**Comment by deathbybandaid** [comment]
well, specifically Locast has the stream limit before they start throttling

**Comment by wouterdebie** [comment]
> well, specifically Locast has the stream limit before they start throttling

I think that's only the case for authenticated and/or API requests and if you redirect, the request will be unauthenticated anyway (i.e the Bearer header will not be set). Secondly, the streams themselves will go to https://hls.locastnet.org, which I don't think is throttled..

**Comment by deathbybandaid** [comment]
Mostly Referencing:

https://helpcenter.locast.org/portal/en/kb/articles/max-streams#:~:text=Max%20Streams%2FRate%20Limit,on%20to%20the%20next%20channel.

**Comment by deathbybandaid** [comment]
Not saying you're wrong about the hls endpoint, but it's still good practice to abide by the terms of a service to avoid a Cease and Desist

**Comment by Croq360** [comment]
The rate limit stuff is part of the reason I set up the DVR as having 3 tuners and not 4.  That way, theoretically, I always have 1 in reserve.

**Comment by Croq360** [comment]
@wouterdebie - I recorded another occurrence of the program that had issues earlier.  It worked fine.  I'll keep an eye out for issues on maybe some other programs and will let you know if/when anything comes up.

**Comment by wouterdebie** [comment]
> Mostly Referencing:
> 
> https://helpcenter.locast.org/portal/en/kb/articles/max-streams#:~:text=Max%20Streams%2FRate%20Limit,on%20to%20the%20next%20channel.

Yeah, but this seems only true for api calls and not for the actual stream. Thinking this through a little: how are you currently limiting the amount of clients? I understand that there is more control when the actual stream flows through your application, but what real mechanisms do you have? The coordination between clients is not there. The only mechanism that currently exists is the fact that Plex/Emby respect the amount of tuners reported when used as a HDHomerun tuner.

**Comment by wouterdebie** [comment]
> The rate limit stuff is part of the reason I set up the DVR as having 3 tuners and not 4. That way, theoretically, I always have 1 in reserve.

True. But this only works for locast2dvr in HDHomerun mode and not for m3u mode, since no tuner count is communicated to Plex/Emby with `tuner.m3u`.

**Comment by deathbybandaid** [comment]
so each origin is assigned X amount of "tuners" (as defined in the config). I then use threading locks to make a tuner "in-use". Per real HDHR docs, a Tuner can be used for scanning as well as streaming.

So if the config for my Locast plugin has 3 tuners, fHDHR controls how many connections to Locast can be made. If no tuners are available, it sends the client the expected response code `response.headers["X-fHDHR-Error"] = "806 - Tune Failed"`.

The whole fHDHR Tuner system was designed to really emulate what the real HDHR box does.

Now that fHDHR supports multiple origins, I can support multiple HDHR interfaces from a single port, and on base url paths


![image](https://user-images.githubusercontent.com/18010545/107816462-cc6e9380-6d42-11eb-9b7d-7a092624d920.png)


Then each origin gets it's own "tuners". 

![image](https://user-images.githubusercontent.com/18010545/107816634-0770c700-6d43-11eb-87dd-01b94009521d.png)


Each origin also has its own channel list, xmltv, and channel editor

![image](https://user-images.githubusercontent.com/18010545/107816728-2ec79400-6d43-11eb-85de-1c1d345298c1.png)

**Comment by deathbybandaid** [comment]
The Tuners are exposed at `ip:port/api/tuners`


https://github.com/fHDHR/fHDHR/blob/main/fHDHR_web/api/tuners.py

**Comment by Croq360** [comment]
> > The rate limit stuff is part of the reason I set up the DVR as having 3 tuners and not 4. That way, theoretically, I always have 1 in reserve.
> 
> True. But this only works for locast2dvr in HDHomerun mode and not for m3u mode, since no tuner count is communicated to Plex/Emby with `tuner.m3u`.

In Emby, when I create an m3u tuner, there's an option for "Simultaneous stream limit".  Wouldn't that accomplish limiting how many things can be watched or recorded at once?

For Plex, since I'm using HDHR mode (I remember reading it doesn't support m3u), I set the number of tuners in the Locast2Dvr config file (well technically, I'm just not overriding the default of 3).  And Plex recognizes it as having 3 tuners.

`--tuner-count COUNT           Tuner count  [default: 3]`

**Comment by wouterdebie** [comment]
@deathbybandaid oh nice! Thanks for the detailed explanation!

**Comment by wouterdebie** [comment]
> In Emby, when I create an m3u tuner, there's an option for "Simultaneous stream limit". Wouldn't that accomplish limiting how many things can be watched or recorded at once?

Yep! That will do the trick! And for HDHR you're right, the tuner count will take care of that.

I do like what @deathbybandaid did, where the amount of used tuners is not only enforced by Plex/Emby, but also by his application. Say that you have multiple Plex/Emby instances, you might want to limit the number of streams at the locast2dvr/fHDHR level. Unfortunately this isn't possible with m3u, because locast2dvr is no longer part of the actual streaming.

**Comment by deathbybandaid** [comment]
@wouterdebie I've had support for m3u for a while, with the m3u url just using the /api/tuners interface. Just today though I added the "passthrough" method like I saw in your code.

https://github.com/fHDHR/fHDHR/blob/main/fHDHR/device/channels/channel.py#L163-L165

**Comment by deathbybandaid** [comment]
In fact, my HDHR plugin adds endpoints that plex/emby expect from a HDHR, and 302's to my /api/tuners

https://github.com/fHDHR/fHDHR_plugin_interface_hdhr/blob/main/web/auto_origin.py

**Comment by wouterdebie** [comment]
@deathbybandaid one thing I wanted to ask you, does SSDP work in your implementation? I've recently disabled it, since neither Plex nor Emby can seem to find locast2dvr. I've tried digging through code and coming up with experiments, but for some reason I can't seem to get it to work. Any pointers?

**Comment by deathbybandaid** [comment]
I created my own SSDP implementation, the one from L2P never worked.

Often in Plex I'll see "A tuner was found" (or whatever the message says) at the top, but I have not been able to get it to show in plex/emby when I try to find them with autodetect.

However, I have a cluster plugin for clustering multiple fHDHR instances together, and that works great. I also can see the SSDP traffic for other devices on my network just fine in the debug logging.

https://github.com/fHDHR/fHDHR/blob/main/fHDHR/device/ssdp/__init__.py

**Comment by deathbybandaid** [comment]
I also added buttons to the web interface to try to manually control SSDP

![image](https://user-images.githubusercontent.com/18010545/107821839-25dac080-6d4b-11eb-9cdc-3cbb282086dd.png)

**Comment by wouterdebie** [comment]
> Often in Plex I'll see "A tuner was found" (or whatever the message says) at the top, but I have not been able to get it to show in plex/emby when I try to find them with autodetect.

Yeah, I've noticed the same thing. 



> However, I have a cluster plugin for clustering multiple fHDHR instances together, and that works great.

How does that work? What do you do with this? And how does it relate to SSDP?



> I also can see the SSDP traffic for other devices on my network just fine in the debug logging.

Yeah, same here.. For some reason, plex/emby doesn't seem to actually download the device.xml or whatever I specify as the location.

**Comment by deathbybandaid** [comment]
Since I was already playing with SSDP, I used SSDP to to detect other fHDHR instances

![image](https://user-images.githubusercontent.com/18010545/107822264-d21ca700-6d4b-11eb-9511-34cf8ac641a7.png)

https://github.com/fHDHR/fHDHR_plugin_interface_cluster

**Comment by deathbybandaid** [comment]
the cluster plugin is less important now that I can do multiple origins from a single instance

**Comment by deathbybandaid** [comment]
![image](https://user-images.githubusercontent.com/18010545/107822420-160fac00-6d4c-11eb-9fdb-209f86157a0b.png)

**Comment by deathbybandaid** [comment]
so when that happened just now, this is all I saw via flask

![image](https://user-images.githubusercontent.com/18010545/107822770-a3530080-6d4c-11eb-8743-7155a8cc5e24.png)


the /device.xml path doesn't even get hit, but how would it have know to look at this path??

**Comment by wouterdebie** [comment]
What is the location that you advertise?

**Comment by deathbybandaid** [comment]
/hdhr/NAMEOFORIGIN/device.xml

I advertise a device.xml per origin

**Comment by wouterdebie** [comment]
> the /device.xml path doesn't even get hit, but how would it have know to look at this path??

Because that's what you advertise, right?

**Comment by deathbybandaid** [comment]
But it doesn't show it being hit in the flask logs

**Comment by wouterdebie** [comment]
I know.. and I think that's why it doesn't show when you're trying to find devices..

**Comment by deathbybandaid** [comment]
I'm toying with a concept to limit client streams for m3u tuning. Instead of redirecting to the Locast M3U8, I'm going to try proxying the m3u8 file, and using some sort of watchdog to determine the client is no longer refreshing

**Comment by Croq360** [comment]
@wouterdebie - Overall, it seems to have been working pretty good with the latest updates.  I did have an issue where a show I had recorded showed the full show when I did a "get info" in Plex, but at particular point while playing back, it jumped about 20 minutes forward in the show.  When I loaded it in VLC, it indicated the show was 20 minutes shorter than in Plex.  So, it appears that the entire show was there, but there's something in the .TS file that is causing the playback to think that section is missing.

I'll save off the log files and try to look at the particular time the show was recording and see if there were any errors during that time.  Maybe the actual data went missing for that 20 minutes somehow.

**Comment by deathbybandaid** [comment]
Made a little progress on my idea, I had a stream method on a seperate thread using `m3u8.load()`, using the `segment.duration` info for how often to refresh, and `playlist.dumps()` into a `io.StringIO()`.

I then redirected the client to an endpoint with access to the proxied m3u8 file. This would also save a `datetime.utcnow` every time the client pulled a fresh copy of the proxied m3u8.

The goal is to then use that timestamp to detect that it hasn't been refreshed in "X" amount of time, so close the tuner.

**Comment by deathbybandaid** [comment]
On the flipside, I added a configurable buffer for streams (defaults to 3 chunks).

![image](https://user-images.githubusercontent.com/18010545/107987022-4436fb00-6f9b-11eb-83c1-857e3e15e882.png)

**Comment by wouterdebie** [comment]
@Croq360 Ok, cool! I found another thing that might be in the way here. Python has a Global Interpreter Lock (GIL), which means that threads don't really run in parallel. If there are multiple streams going on that do quite a bit of I/O, it might be that locast2dvr can't keep up, since the whole process only runs on a single core and that could theoretically cause things to slow down. I'm working on forking different processes for when locast2dvr consumes a lot of data that is time critica (so watch with ffmpeg and watch_direct). I'll let you know when I have an update.

**Comment by wouterdebie** [comment]
@deathbybandaid very cool approach! You base the client id on IP and source port?

**Comment by wouterdebie** [comment]
@deathbybandaid something else I figured out is that HDHomerun devices use 65001/udp for discovery besides SSDP. I have discovery working with a test UDP server and plex, but because of https://emby.media/community/index.php?/topic/94327-hdhomerun-detect-my-devices-bug-using-wrong-hostnameport/&tab=comments#comment-985946, this only works when you run your HDHomerun device on port 80.

**Comment by deathbybandaid** [comment]
@wouterdebie yeah, research the other day suggested mdns being a key factor as well. In theory an nginx reverse proxy might help with port 80 issues

I assign clients a uuid at connection to flask.

**Comment by deathbybandaid** [comment]
I'm interested in what you find out for that port. I'll add whatever it is to my hdhr plugin when we nail it. I'm also working with @DanAustinGH (who has a real HDHR) and wireshsrk to try and capture some data.

**Comment by deathbybandaid** [comment]
@wouterdebie what's all this new UUID/Checksum stuff you've got going on?

**Comment by wouterdebie** [comment]
@deathbybandaid I'm trying to get as close to a libhomerun implementation as possible. So I'm generating actual uuids and making sure the homerun device is valid: https://github.com/Silicondust/libhdhomerun/blob/7c54382fb681d03888b469033e50bebaf4ce6bce/hdhomerun_discover.c#L528

**Comment by wouterdebie** [comment]
@deathbybandaid Working 65001/udp server: https://gist.github.com/wouterdebie/f8af08b1157659eb0f3c5286d2e4d012

**Comment by deathbybandaid** [comment]
thanks for the info, I'm working on how to implement the uuid stuff into fHDHR_plugin_interface_hdhr.

I also added the ability for interface plugins to have a function to start threads, so that's what I'll do for this new UDP port

**Comment by deathbybandaid** [comment]
Got the UID stuff implemented. Really interesting

![image](https://user-images.githubusercontent.com/18010545/108084993-d8a86880-7042-11eb-9ba9-8dc3b9457a1a.png)

**Comment by deathbybandaid** [comment]
That UDP service was interesting to implement. Still had to switch to port 80 for emby to see stuff

![image](https://user-images.githubusercontent.com/18010545/108095357-c4b63400-704d-11eb-9180-7db1e48b5e4a.png)


I even managed to do a for-loop over my origins and send the data

https://github.com/fHDHR/fHDHR_plugin_interface_hdhr/pull/9

**Comment by deathbybandaid** [comment]
Still has some work to do

**Comment by deathbybandaid** [comment]
I'm thinking playing with mdns might make it so I can direct to the correct origin

**Comment by deathbybandaid** [comment]
also I side note, xteve is on a non-80 port and shows up in Plex

**Comment by deathbybandaid** [comment]
@wouterdebie https://community.getchannels.com/t/channels-with-hrtunerproxy/3841/4

I think 65001 is also a TCP binding that listens as well

**Comment by deathbybandaid** [comment]
https://github.com/geertbongers/tvhProxy/commit/f3c77827f92892d0a1d3e35a1d523aaffeb8f8e6#diff-261d534dab77c66913c49ddea5b38deea49a9b1c40c5b7c095b50192637dc729R109-R140

**Comment by deathbybandaid** [comment]
did some testing with the TCP method, and didn't see any traffic or logging with plex or emby

**Comment by deathbybandaid** [comment]
restarted my plex server, and now it's detecting fHDHR at a non-80 port, but still not respecting the URL path (just like emby).

**Comment by deathbybandaid** [comment]
new issue, and an odd one,,,

TvHeadend  is crashing when it detects the "hdhr" udp device on the network

**Comment by deathbybandaid** [comment]
really interesting, I can disable the UDP service in fHDHR via my config, and tvheadend will run fine.

Turn it back on and 

````
CRASH: Fault address (nil) (Address not mapped)
````

**Comment by deathbybandaid** [comment]
I also want to note a user of fHDHR had trouble binding on Windows

**Comment by Croq360** [comment]
So, I haven't had any more gaps in videos.  Things seem pretty solid in PMS for several days now.

**Comment by wouterdebie** [comment]
Great! I'm actually rewriting the whole thing in Rust, which should be way
more performant! More on this topic soon!

**Comment by wouterdebie** [comment]
I just pushed a workin Rust version at https://github.com/wouterdebie/locast2tuner. Someone might want to give this a try :)

**Comment by Croq360** [comment]
I saw that just now.  Giving it a try.  One initial observation is that the
config file needs 'true' for flags instead of the "1" that was in the
Python version.  For example,

multiplex=true

I'll let you know how it goes.

**Comment by wouterdebie** [comment]
Yes there are a few difference here and there. Let me know how it goes.

**Comment by wouterdebie** [comment]
@deathbybandaid Not sure if you've seen it, but I've fully moved over to locast2tuner development. With rust I've been able to get nice packages and such :) Apart from that Rust is awesome!
Did you ever figure SSDP out? I have not been able to get Plex or Emby to discover anything and I'm thinking of just not implementing this at all.

**Comment by deathbybandaid** [comment]
It's easy enough to copy ip:port into emby/plex manually anyway.
